### PR TITLE
Detect Sapphire Rapids features

### PR DESCRIPTION
**Pull request by askervin**
This patch depends on PR #59.

### CODE REVIEW

**Comment by askervin** [comment]
@mythi , detecting `HYBRID` CPU dropped as it is not a Sapphire Rapids feature. And It could use some better way to expose detected hybrid cores, anyway.

**Comment by askervin** [comment]
Rebased on top of both commits in PR #59.

**Comment by askervin** [comment]
@mythi , @klauspost , tested on Sapphire Rapids with PR #59. Works fine. :)

**Comment by klauspost** [comment]
Thanks! I will see if I can squeeze it in

**Comment by mythi** [comment]
@klauspost that'd be great! We're using `cpuid` in kubernetes node-feature-discovery to detect cluster node CPU features and would like to get a new release out with these changes.

File: cpuid.go [hunk 1]
```diff
@@ -1146,81 +1141,82 @@ func hasSGX(available, lc bool) (rval SGXSupport) {
...
 	mfi := maxFunctionID()
 	vend, _ := vendorID()
 	if mfi < 0x1 {
-		return 0, 0
+		return 0, nil
 	}
-	flags := uint64(0)
-	amxFlags := AmxFlags(0)
+
+	fs := make(FlagSet, LASTID/64+1)
```
Comment (marquiz) [inline cpuid.go:1208]: Would it be safer to have a function for doing this, e.g. `newFlagSet()` or smth

File: cpuid.go [hunk 2]
```diff
@@ -231,27 +241,14 @@ var flagNamesArm = map[ArmFlags]string{
...
-}
-
 // CPUInfo contains information about the detected system CPU.
 type CPUInfo struct {
 	BrandName      string   // Brand name reported by the CPU
 	VendorID       Vendor   // Comparable CPU vendor ID
 	VendorString   string   // Raw vendor string.
-	Features       Flags    // Features of the CPU (x64)
+	Features       Flags    // Features of the CPU (x64) (deprecated)
+	flagSet        FlagSet  // Features of the CPU
```
Comment (marquiz) [inline cpuid.go:272]: I was a bit distracted by the name `flagSet` here as it's overloaded with the `flags` package. What do you think about renaming it? `features2` or `featureSet`, for example((?)

**Comment by marquiz** [comment]
Two small comments but otherwise looks good to me

**Comment by marquiz** [comment]
ping @klauspost , any comments(?) We'd really like to get this merged 😊

**Comment by mythi** [comment]
@askervin @klauspost: I've pushed v2 to #59. Changes: 1) changed `Flags()`/`flagSet` to `FeatureSet()/featureSet`, 2) added a Test case for "compatibility" to make sure the backwards compatibility to `CPU.Features` stringer is kept.

**Comment by askervin** [comment]
Ported and rebased on top of @mythi 's latest commits.

File: cpuid.go (hunk 1 above)
Comment (askervin) [reply cpuid.go:1221]: This is in PR #59, closing in this conversation.

File: cpuid.go (hunk 2 above)
Comment (askervin) [reply cpuid.go]: This is in PR #59. Fixed there in the latest update. (This was well spotted, naming looks better now!)

**Comment by askervin** [comment]
@mythi, @klauspost , rebased on top of mythi's latest commits. Good to go.

**Comment by klauspost** [comment]
@askervin #59 is now merged.

**Comment by askervin** [comment]
@klauspost , great, thank you! I'll rebase this on top of master.

**Comment by askervin** [comment]
Done!

**Comment by klauspost** [comment]
@askervin Thanks. I will review the package and prepare a v2 release.

**Pull request merged**